# In this file, we train a model, Random Forest Regressor, prior to the feature reduction process to test how well the model can predict the target.  This does not include every column from the raw datasets but will, nevertheless, include every column that remains following the basic preliminary cleaning stage.  Thus, we will import and merge each of the 3 cleaned data tables from postgres to perform this "baseline" testing procedure.

In [20]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2
import time

In [21]:
from pathlib import Path
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [22]:
# Download raw data from postgres for stage 1 ETL

conn_string = 'postgres://whnpmxwsiccrtg:53c453893549d2b1e6a4ff92e626a2a08ebcaff66678e50d33e3742f66e3e4f4@ec2-52-4-171-132.compute-1.amazonaws.com/d2ajro4cjr10lb'

db = create_engine(conn_string)
conn = db.connect()

start_time = time.time()
clean_listing = pd.read_sql_query('select * from "listings_full_clean"',con=conn)
calendar = pd.read_sql_query('select * from "calendar_clean"',con=conn)
amenities = pd.read_sql_query('select * from "amenities_clean"',con=conn)
print("PostGres Download Duration: {} seconds".format(time.time() - start_time))


PostGres Download Duration: 1.8744537830352783 seconds


In [23]:
# View clean_listing database
clean_listing.head()

id last_scraped host_since  host_listings_count host_is_superhost  \
0   6422   2019-01-17 2009-04-03                    1                 t   
1  20847   2019-01-17 2010-02-08                    2                 t   
2  25341   2019-01-17 2010-02-08                    2                 t   
3  25613   2019-01-17 2010-03-18                    1                 t   
4  37306   2019-01-17 2010-07-08                    1                 f   

  host_identity_verified neighbourhood_cleansed   latitude  longitude  \
0                      t             District 6  36.173151 -86.735813   
1                      f             District 7  36.196397 -86.704267   
2                      f             District 7  36.197053 -86.704937   
3                      t             District 6  36.178263 -86.741622   
4                      f            District 17  36.133109 -86.774767   

      room_type property_type  accommodates  bathrooms  bedrooms  bed_type  \
0  Private room         House             2        1.0         2  Real Bed   
1  Private room         House             2        1.0         1  Real Bed   
2  Private room         House             2        1.0         1  Real Bed   
3  Private room         House             2        1.0         1  Real Bed   
4  Private room         House             2        1.0         1  Real Bed   

   price  security_deposit  cleaning_fee  review_scores_rating  \
0   40.0               0.0           0.0                  99.0   
1   41.0               0.0          25.0                  94.0   
2   65.0               0.0          25.0                  96.0   
3   93.0               0.0           0.0                  97.0   
4  105.0             200.0           0.0                  99.0   

   number_of_reviews  review_scores_cleanliness  review_scores_location  \
0                561                       10.0                    10.0   
1                138                       10.0                     9.0   
2                102                       10.0                     9.0   
3                454                       10.0                    10.0   
4                100                       10.0                    10.0   

   review_scores_communication  review_scores_checkin  review_scores_value  \
0                         10.0                   10.0                 10.0   
1                         10.0                    9.0                  9.0   
2                         10.0                   10.0                  9.0   
3                         10.0                   10.0                 10.0   
4                         10.0                   10.0                 10.0   

  instant_bookable is_business_travel_ready          cancellation_policy  \
0                f                        f  strict_14_with_grace_period   
1                f                        f  strict_14_with_grace_period   
2                f                        f  strict_14_with_grace_period   
3                t                        f                     moderate   
4                f                        f  strict_14_with_grace_period   

  require_guest_profile_picture require_guest_phone_verification  \
0                             f                                f   
1                             t                                t   
2                             t                                t   
3                             t                                t   
4                             t                                t   

  has_availability  is_review  days_host  
0                t       True       3576  
1                t       True       3265  
2                t       True       3265  
3                t       True       3227  
4                t       True       3115

In [24]:
# View data types of full_listings
clean_listing.dtypes

id                                           int64
last_scraped                        datetime64[ns]
host_since                          datetime64[ns]
host_listings_count                          int64
host_is_superhost                           object
host_identity_verified                      object
neighbourhood_cleansed                      object
latitude                                   float64
longitude                                  float64
room_type                                   object
property_type                               object
accommodates                                 int64
bathrooms                                  float64
bedrooms                                     int64
bed_type                                    object
price                                      float64
security_deposit                           float64
cleaning_fee                               float64
review_scores_rating                       float64
number_of_reviews              

In [25]:
# Drop duplicated variables.
listing = clean_listing.drop(columns = ['last_scraped', 'host_since', 'latitude', 'longitude', 'price'])

In [26]:
# View amenities dataframe
amenities.head()

id   tv  internet  wifi  air_conditioning  kitchen  \
0   6422  1.0       1.0   1.0               1.0      1.0   
1  20847  1.0       0.0   1.0               1.0      1.0   
2  25341  1.0       0.0   1.0               1.0      1.0   
3  25613  1.0       1.0   1.0               1.0      1.0   
4  37306  1.0       0.0   1.0               1.0      1.0   

   free_parking_on_premises  pets_live_on_this_property  dogs  cats  heating  \
0                       1.0                         1.0   1.0   1.0      1.0   
1                       1.0                         0.0   0.0   0.0      1.0   
2                       1.0                         0.0   0.0   0.0      1.0   
3                       1.0                         0.0   0.0   0.0      1.0   
4                       0.0                         1.0   1.0   1.0      1.0   

   family/kid_friendly  washer  dryer  smoke_detector  \
0                  1.0     1.0    1.0             1.0   
1                  0.0     1.0    1.0             1.0   
2                  0.0     1.0    1.0             1.0   
3                  0.0     1.0    1.0             1.0   
4                  0.0     1.0    1.0             1.0   

   carbon_monoxide_detector  first_aid_kit  safety_card  fire_extinguisher  \
0                       1.0            1.0          1.0                1.0   
1                       1.0            1.0          0.0                1.0   
2                       1.0            1.0          0.0                1.0   
3                       0.0            1.0          0.0                1.0   
4                       0.0            0.0          0.0                0.0   

   essentials  shampoo  lock_on_bedroom_door  hangers  hair_dryer  iron  \
0         1.0      1.0                   1.0      1.0         1.0   1.0   
1         1.0      1.0                   1.0      1.0         1.0   1.0   
2         1.0      1.0                   1.0      1.0         1.0   1.0   
3         1.0      1.0                   0.0      1.0         1.0   1.0   
4         1.0      1.0                   1.0      1.0         1.0   1.0   

   laptop_friendly_workspace  bathtub  children’s_books_and_toys  hot_water  \
0                        1.0      1.0                        1.0        1.0   
1                        1.0      0.0                        0.0        1.0   
2                        1.0      0.0                        0.0        1.0   
3                        1.0      0.0                        0.0        0.0   
4                        1.0      0.0                        0.0        1.0   

   bed_linens  extra_pillows_and_blankets  ethernet_connection  microwave  \
0         1.0                         1.0                  1.0        1.0   
1         1.0                         1.0                  0.0        0.0   
2         1.0                         1.0                  0.0        1.0   
3         0.0                         0.0                  0.0        0.0   
4         0.0                         0.0                  0.0        0.0   

   coffee_maker  refrigerator  dishes_and_silverware  garden_or_backyard  \
0           1.0           1.0                    1.0                 1.0   
1           0.0           0.0                    0.0                 0.0   
2           1.0           1.0                    1.0                 0.0   
3           0.0           0.0                    0.0                 0.0   
4           0.0           0.0                    0.0                 0.0   

   luggage_dropoff_allowed  long_term_stays_allowed  host_greets_you  \
0                      1.0                      1.0              1.0   
1                      1.0                      1.0              0.0   
2                      1.0                      1.0              0.0   
3                      0.0                      0.0              1.0   
4                      1.0                      0.0              1.0   

   cable_tv  breakfast  free_street_parking  indoor_fireplace  \
0       

In [27]:
# View calendar dataframe
calendar.head(10)

id      day  month  price
0  6422  weekday      1   40.0
1  6422  weekday      2   40.0
2  6422  weekday      3   40.0
3  6422  weekday      4   40.0
4  6422  weekday      5   40.0
5  6422  weekday      6   40.0
6  6422  weekday      7   40.0
7  6422  weekday      8   40.0
8  6422  weekday      9   40.0
9  6422  weekday     10   40.0

In [28]:
# Merge amenities table to full listings.
merge1 = listing.merge(amenities, how='left', on ='id')

In [29]:
# Merge calendar table to amenities and full listings.
merged = merge1.merge(calendar, how='left', on ='id')
merged.head(20)

0.0             0.0             0.0                       0.0   
14              0.0             0.0             0.0                       0.0   
15              0.0             0.0             0.0                       0.0   
16              0.0             0.0             0.0                       0.0   
17              0.0             0.0             0.0                       0.0   
18              0.0             0.0             0.0                       0.0   
19              0.0             0.0             0.0                       0.0   

    firm_mattress  baby_bath  changing_table  hot_water_kettle  \
0             0.0        0.0             0.0               0.0   
1             0.0        0.0             0.0               0.0   
2             0.0        0.0             0.0               0.0   
3             0.0        0.0             0.0               0.0   
4             0.0        0.0             0.0               0.0   
5             0.0        0.0             0.0               0.0   
6             0.0        0.0             0.0               0.0   
7             0.0        0.0             0.0               0.0   
8             0.0        0.0             0.0               0.0   
9             0.0        0.0             0.0               0.0   
10            0.0        0.0             0.0               0.0   
11            0.0        0.0             0.0               0.0   
12            0.0        0.0             0.0               0.0   
13            0.0        0.0             0.0               0.0   
14            0.0        0.0             0.0               0.0   
15            0.0        0.0             0.0               0.0   
16            0.0        0.0             0.0               0.0   
17            0.0        0.0             0.0               0.0   
18            0.0        0.0             0.0               0.0   
19            0.0        0.0             0.0               0.0   

    central_air_conditioning  breakfast_table  murphy_bed  outdoor_parking  \
0                        0.0              0.0         0.0              0.0   
1                        0.0              0.0         0.0              0.0   
2                        0.0              0.0         0.0              0.0   
3                        0.0              0.0         0.0              0.0   
4                        0.0              0.0         0.0              0.0   
5                        0.0              0.0         0.0              0.0   
6                        0.0              0.0         0.0              0.0   
7                        0.0              0.0         0.0              0.0   
8                        0.0              0.0         0.0              0.0   
9                        0.0              0.0         0.0              0.0   
10                       0.0              0.0         0.0              0.0   
11                       0.0              0.0         0.0              0.0   
12                       0.0              0.0         0.0              0.0   
13                       0.0              0.0         0.0              0.0   
14                       0.0              0.0         0.0              0.0   
15                       0.0              0.0         0.0              0.0   
16                       0.0              0.0         0.0              0.0   
17                       0.0              0.0         0.0              0.0   
18                       0.0              0.0         0.0              0.0   
19                       0.0              0.0         0.0              0.0   

    ev_charger  window_guards  shower_chair  stair_gates  game_console  \
0          0.0            0.0           0.0          0.0           0.0   
1          0.0            0.0           0.0          0.0           0.0   
2          0.0            0.0           0.0          0.0           0.0   
3          0.0            0.0           0.0          0.0           0.0   
4          0.0            0.0           0.0         

In [30]:
# Merge tables and drop 'id' (unique identifier - not relevant)
merged = merged.drop(columns = 'id')

In [31]:
# Create a copy of the database to work with going forward.
merged2 = merged.copy()

In [32]:
# Create a list of the object type columns.
objects = merged2.dtypes[merged2.dtypes == 'object'].index.tolist()
objects

['host_is_superhost',
 'host_identity_verified',
 'neighbourhood_cleansed',
 'room_type',
 'property_type',
 'bed_type',
 'instant_bookable',
 'is_business_travel_ready',
 'cancellation_policy',
 'require_guest_profile_picture',
 'require_guest_phone_verification',
 'has_availability',
 'day']

In [33]:
# Create a OneHotEncoder instance
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(sparse=False, drop='if_binary')

# Fit and transform the OneHotEncoder using the categorical variable list
encode_df = pd.DataFrame(enc.fit_transform(merged2[objects]))

# Add the encoded variable names to the dataframe
encode_df.columns = enc.get_feature_names(objects)
encode_df.head()

host_is_superhost_t  host_identity_verified_t  \
0                  1.0                       1.0   
1                  1.0                       1.0   
2                  1.0                       1.0   
3                  1.0                       1.0   
4                  1.0                       1.0   

   neighbourhood_cleansed_District 1  neighbourhood_cleansed_District 10  \
0                                0.0                                 0.0   
1                                0.0                                 0.0   
2                                0.0                                 0.0   
3                                0.0                                 0.0   
4                                0.0                                 0.0   

   neighbourhood_cleansed_District 11  neighbourhood_cleansed_District 12  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 13  neighbourhood_cleansed_District 14  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 15  neighbourhood_cleansed_District 16  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 17  neighbourhood_cleansed_District 18  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 19  neighbourhood_cleansed_District 2  \
0                                 0.0                                0.0   
1                                 0.0                                0.0   
2                                 0.0                                0.0   
3                                 0.0                                0.0   
4                                 0.0                                0.0   

   neighbourhood_cleansed_District 20  neighbourhood_cleansed_District 21  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

   neighbourhood_cleansed_District 22  neighbourhood_cleansed_District 23  \
0                                 0.0                                 0.0   
1                                 0.0                                 0.0   
2                                 0.0                                 0.0   
3                                 0.0                                 0.0   
4                                 0.0                                 0.0   

In [34]:
# Merge one-hot encoded features and drop the originals
merged2 = merged2.merge(encode_df,left_index=True, right_index=True)
merged2 = merged2.drop(columns=objects)

In [35]:
merged2.head()

0           0.0   
1                         0.0                       0.0           0.0   
2                         0.0                       0.0           0.0   
3                         0.0                       0.0           0.0   
4                         0.0                       0.0           0.0   

   gas_oven  formal_dining_area  beach_essentials  table_corner_guards  \
0       0.0                 0.0               0.0                  0.0   
1       0.0                 0.0               0.0                  0.0   
2       0.0                 0.0               0.0                  0.0   
3       0.0                 0.0               0.0                  0.0   
4       0.0                 0.0               0.0                  0.0   

   air_purifier  rain_shower  bathtub_with_bath_chair  day_bed  \
0           0.0          0.0                      0.0      0.0   
1           0.0          0.0                      0.0      0.0   
2           0.0          0.0                      0.0      0.0   
3           0.0          0.0                      0.0      0.0   
4           0.0          0.0                      0.0      0.0   

   pillow-top_mattress  ground_floor_access  dvd_player  warming_drawer  \
0                  0.0                  0.0         0.0             0.0   
1                  0.0                  0.0         0.0             0.0   
2                  0.0                  0.0         0.0             0.0   
3                  0.0                  0.0         0.0             0.0   
4                  0.0                  0.0         0.0             0.0   

   printer  mudroom  amazon_echo  shared_pool  hbo_go  espresso_machine  \
0      0.0      0.0          0.0          0.0     0.0               0.0   
1      0.0      0.0          0.0          0.0     0.0               0.0   
2      0.0      0.0          0.0          0.0     0.0               0.0   
3      0.0      0.0          0.0          0.0     0.0               0.0   
4      0.0      0.0          0.0          0.0     0.0               0.0   

   heated_floors  alfresco_bathtub  steam_oven  double_oven  wine_cooler  \
0            0.0               0.0         0.0          0.0          0.0   
1            0.0               0.0         0.0          0.0          0.0   
2            0.0               0.0         0.0          0.0          0.0   
3            0.0               0.0         0.0          0.0          0.0   
4            0.0               0.0         0.0          0.0          0.0   

   soaking_tub  beachfront  ski-in/ski-out  beach_view  \
0          0.0         0.0             0.0         0.0   
1          0.0         0.0             0.0         0.0   
2          0.0         0.0             0.0         0.0   
3          0.0         0.0             0.0         0.0   
4          0.0         0.0             0.0         0.0   

   high-resolution_computer_monitor  standing_valet  pool_with_pool_hoist  \
0                               0.0             0.0                   0.0   
1                               0.0             0.0                   0.0   
2                               0.0             0.0                   0.0   
3                               0.0             0.0                   0.0   
4                               0.0             0.0                   0.0   

   sun_loungers  electric_profiling_bed  hammock  mobile_hoist  month  price  \
0           0.0                     0.0      0.0           0.0      1   40.0   
1           0.0                     0.0      0.0           0.0      2   40.0   
2           0.0                     0.0      0.0           0.0      3   40.0   
3           0.0                     0.0      0.0           0.0      4   40.0   
4           0.0                     0.0      0.0           0.0      5   40.0   

   host_is_superhost_t  host_identity_verified_t  \
0                  1.0                       1.0   
1                  1.0                       1.0   
2                  1.0                

In [36]:
pd.pandas.set_option('display.max_columns', None)

In [37]:
# Define the target variable.
y=merged2['price']

In [40]:
# Define the features.
X = merged2.drop(columns ='price')
# X = X.values.reshape(-1, 201)

In [41]:
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [44]:
# import the regressor 
from sklearn.ensemble import RandomForestRegressor

#Instantiate the Decision Tree Regression model:
random_forest_regression_model = RandomForestRegressor(n_estimators = 100, random_state = 78) 
random_forest_regression_model.fit(X_train, y_train)

In [ ]:
# Calculate the r-squared score of the initial model.
random_forest_regression_model.score(X_test, y_test, sample_weight=None)

# From the above calculation, our random forest regressor produced an r-squared value of 95%

# Now we will calculate the mean squared error:

In [23]:
# Predict using the Decision Tree Random Forest Regression model.
y_pred = random_forest_regression_model.predict(X_test)
y_pred

array([445.69152833, 238.16875   , 461.5385    , ..., 176.08095238,
       226.85513889,  46.72571429])

In [24]:
#Calculate the mean squared error
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
rmse

53.861577220224966

# From the above calculation, our Random Forest Regressor was able to predict our continous target (price) with a mean squared error value of 53.86.  This indicates our model is off, on average, by about $7.34 (since the square root of 53.86 is about 7.34)

In [ ]:
# adjusted r2 score
# For loop to start at all features and loop to decrease features each loop to optimize adjusted r2 score